### PR TITLE
PromQL engine refactoring

### LINKED ISSUE #6739: Remove LookbackDelta and DefaultEvaluationInterval from the global variables of promql
## Proposal
**Use case. Why is this important?**

The behaviour of a PromQL engine should not depend on global variables of the PromQL package, especially when we alter those global variables at runtime.

https://github.com/prometheus/prometheus/blob/0e912faf4f52ed3f803cc18d0ebbd4eb0ba5a8bd/promql/engine.go#L55-L63

Therefore I would like to move them as EngineOpts

https://github.com/prometheus/prometheus/blob/0e912faf4f52ed3f803cc18d0ebbd4eb0ba5a8bd/promql/engine.go#L216-L223

### PR DESCRIPTION
**Pull request by roidelapluie**
## Change 1

Following https://golang.org/pkg/context/:
 Do not store Contexts inside a struct type; instead, pass a Context
explicitly to each function that needs it. The Context should be the
first parameter, typically named ctx.

I have seen no obvious reason not to do so in the engine.

## Change 2 and 3

Remove LookbackDelta and DefaultEvaluationInterval from the global variables of promql

closes  #6739 

Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

### CODE REVIEW

**Comment by brian-brazil** [comment]
Do you know why it's suggested not to store it in a struct? We use one struct per query, so I don't see a problem with what we're currently doing.

**Comment by roidelapluie** [comment]
I guess that they do not want a context to cancel a parent function's context.

**Comment by roidelapluie** [comment]
Also I would like to see how we could implement time out when reading TSDB. Currently that part of ther query can never time out, in such a situation we want to use context correctly to start with.

**Comment by roidelapluie** [comment]
> https://github.com/prometheus/prometheus/issues/5878

File: promql/engine.go [hunk 1]
```diff
@@ -350,6 +340,21 @@ func (ng *Engine) SetQueryLogger(l QueryLogger) {
 	}
 }
 
+// GetLookbackDelta returns the lookback delta of the query engine.
+func (ng *Engine) GetLookbackDelta() time.Duration {
```
Comment (juliusv) [inline promql/engine.go]: Plain getters that don't require any mutexes / special atomic code etc. are unusual in Go - instead, just exporting the struct field is more common / preferred, unless there's a special reason to abstract it behind a function.

**Comment by juliusv** [comment]
For contexts, despite that advice I think it's totally fine to keep them in struct fields, as long as those structs are request/query-scoped, or whatever the scope of that context should be.

**Comment by roidelapluie** [comment]
looks like there is no real concensus now

**Comment by juliusv** [comment]
@roidelapluie FWIW I liked the bits that got rid of global variables, but would maybe be best to submit that separately.

**Comment by roidelapluie** [comment]
@juliusv TBH LookbackDelta should really go away from global because we pass it in the evaluator struct anyway. Will open a PR.

**Comment by juliusv** [comment]
@roidelapluie Since it should almost never be changed, how about keeping a global `const DefaultLookbackDelta` (instead of `var`) and then if a user leaves the `LookbackDelta` to its zero value in the options struct, set it to the default value?

**Comment by roidelapluie** [comment]
> @roidelapluie Since it should almost never be changed, how about keeping a global `const DefaultLookbackDelta` (instead of `var`) and then if a user leaves the `LookbackDelta` to its zero value in the options struct, set it to the default value?

Why would we remove the ability to disable the loopback delta?

**Comment by juliusv** [comment]
@roidelapluie Do you mean you actually want users to be able to set it to `0` and make vector selectors not select anything?

**Comment by roidelapluie** [comment]
It would select if the timestamp is right

**Comment by juliusv** [comment]
Hmmm yeah... I wonder if there's a use case for that. But maybe still good to not rule it out for now, and always make the setting mandatory then. However, it might unknowingly break existing users if they just update the PromQL package and aren't aware that they need to set a new field in the options or otherwise get no lookback :)

**Comment by roidelapluie** [comment]
> Hmmm yeah... I wonder if there's a use case for that. But maybe still good to not rule it out for now, and always make the setting mandatory then. However, it might unknowingly break existing users if they just update the PromQL package and aren't aware that they need to set a new field in the options or otherwise get no lookback :)

let's continue in #6746 . I have used a pointer to check if the value is defined

**Pull request closed without merging**